### PR TITLE
Fetch AN amendements in batches

### PR DESCRIPTION
**Pull request by ronnix**

### CODE REVIEW

File: repondeur/zam_repondeur/tasks/fetch.py [hunk 1]
```diff
@@ -101,50 +104,75 @@ def fetch_amendements(lecture_pk: Optional[int]) -> bool:
...
-        # on some interactive operations).
+        nb_batches = 1
+        start_index = 0
 
-        # But during tests we want everything to run in a single transaction that
-        # we can roll back at the end.
-        if not huey.immediate:
-            transaction.commit()
-            DBSession.expire_all()
+        while True:
```
Comment (davidbgk) [inline repondeur/zam_repondeur/tasks/fetch.py]: Given that we have a `total_timer` maybe we can add a timeout to that loop if it takes too much time?

File: repondeur/zam_repondeur/services/fetch/an/amendements.py [hunk 2]
```diff
@@ -181,14 +200,16 @@ def _collect_amendements_other(
...
             except Exception:
-                logger.exception("Error while fetching amendement %r", numero)
-                errored.append(str(numero))
+                logger.exception("Error while fetching amendement %r", numero_prefixe)
+                errored.append(numero_prefixe)
                 continue
-            if numero > max_num_seen:
-                max_num_seen = numero
-        return actions, unchanged, errored
+        return actions, unchanged, errored, nb_404
```
Comment (davidbgk) [inline repondeur/zam_repondeur/services/fetch/an/amendements.py]: At that point, we might want to return a structure? 🤔

File: repondeur/zam_repondeur/services/fetch/an/amendements.py [hunk 3]
```diff
@@ -57,10 +61,26 @@ def __init__(self, organe: str) -> None:
...
+    def advance(self, offset: int) -> None:
+        self.lecture.set_fetch_progress(self.start_index + offset + 1, self.total)
+
+
 class AssembleeNationale(RemoteSource):
     def __init__(self, settings: Dict[str, Any], prefetching_enabled: bool = True):
         super().__init__(settings=settings, prefetching_enabled=prefetching_enabled)
+        self.batch_size = int(settings.get("zam.fetch.an.batch_size", BATCH_SIZE))
         self.max_404 = int(settings.get("zam.fetch.an.max_404", MAX_404))
+        if self.max_404 > self.batch_size:
```
Comment (davidbgk) [inline repondeur/zam_repondeur/services/fetch/an/amendements.py]: I wonder if we can make that try at launch time?

File: repondeur/zam_repondeur/services/fetch/an/amendements.py (hunk 2 above)
Comment (davidbgk) [reply repondeur/zam_repondeur/services/fetch/an/amendements.py]: Mmmh re-taking a look at the real usage I'm not sure anymore…

File: repondeur/zam_repondeur/services/fetch/an/amendements.py [hunk 4]
```diff
@@ -461,23 +480,56 @@ def build_url(
...
     """
-    Data extaction for Assemblée Nationale dérouleur
+    Data extraction for Assemblée Nationale dérouleur
     """
 
     def __init__(self, lecture: Lecture, content: dict):
         self.lecture = lecture
-        self.content = content
+        self.discussion_items = self._discussion_items(content)
+        self.position = self._position(self.discussion_items)
```
Comment (davidbgk) [inline repondeur/zam_repondeur/services/fetch/an/amendements.py]: Why not `positionS`? If I understand correctly it stores a dict with all the positions?

Comment (ronnix) [reply repondeur/zam_repondeur/services/fetch/an/amendements.py]: I refactored all this part

File: repondeur/zam_repondeur/services/fetch/an/amendements.py [hunk 5]
```diff
@@ -57,10 +63,27 @@ def __init__(self, organe: str) -> None:
         self.organe = organe
 
 
+class ProgressBar:
```
Comment (davidbgk) [inline repondeur/zam_repondeur/services/fetch/an/amendements.py]: I wonder if we should/could reuse the [ProgressBar from progressist](https://github.com/pyrates/progressist/blob/master/progressist/__init__.py#L47) (we already have that dependency) 🤔

Comment (ronnix) [reply repondeur/zam_repondeur/services/fetch/an/amendements.py]: No idea

File: repondeur/zam_repondeur/tasks/fetch.py (hunk 1 above)
Comment (ronnix) [reply repondeur/zam_repondeur/tasks/fetch.py]: How long is too long?

Comment (davidbgk) [reply repondeur/zam_repondeur/tasks/fetch.py]: 100000/250*120/60/60 => 13h20 if we consider a max loop at 2 minutes (120 seconds) 😰

**Pull request merged**